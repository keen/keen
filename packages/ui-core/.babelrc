{
  "presets": [
    ["@babel/preset-env", { "modules": false }],
  ],
  "plugins": [
    [
      "@quickbaseoss/babel-plugin-styled-components-css-namespace",
      {"cssNamespace": "&&&"}
    ],
    ["babel-plugin-styled-components", {
      "namespace": "keen-ui-core",
      "minify": true,
      "transpileTemplateLiterals": true
    }]
  ],
}
