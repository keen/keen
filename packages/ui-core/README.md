# `@keen.io/ui-core`

[![Known Vulnerabilities](https://snyk.io/test/github/keen/keen/badge.svg?targetFile=packages/ui-core/package.json)](https://snyk.io/test/github/keen/keen?targetFile=packages/ui-core/package.json)

Set of reusable visual `React` components used in `keen.io` Front-End projects.

## Build

The `@keen.io/ui-core` use two step build to address issues with CSS specificity.

##### Typescript

First stage is responsible for transpilation Typescript code to `esnext` and emiting type declarations.

##### Babel

Second stage is responsible for code transpilation based on supported `.browserslistrc` file and increasing `styled-components` css specificity by using plugins from `.babelrc` file.

### npm scripts

List of useful commands that could be used by developers. Execution in the command-line interface should be prefixed with `yarn` package manager.

| Command | Description                      |
| ------- | -------------------------------- |
| `test`  | run unit tests.                  |
| `build` | builds application distribution. |
