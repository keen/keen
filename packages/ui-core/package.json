{
  "name": "@keen.io/ui-core",
  "version": "2.40.2",
  "description": "Keen visual components",
  "author": "Keen.IO <team@keen.io> (https://keen.io/)",
  "homepage": "https://github.com/keen/keen#readme",
  "license": "MIT",
  "main": "dist/index.js",
  "types": "typings/index.d.ts",
  "module": "src/index.ts",
  "directories": {
    "lib": "dist"
  },
  "files": [
    "dist",
    "typings"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/keen/keen.git"
  },
  "scripts": {
    "test": "jest",
    "prebuild": "rimraf dist lib tsconfig.tsbuildinfo",
    "build": "tsc",
    "postbuild": "babel lib --out-dir dist"
  },
  "bugs": {
    "url": "https://github.com/keen/keen/issues"
  },
  "peerDependencies": {
    "moment-timezone": ">=0.5.30",
    "react": ">=16",
    "react-dom": ">=16",
    "styled-components": ">=5"
  },
  "dependencies": {
    "@keen.io/charts": "^2.31.2",
    "@keen.io/charts-utils": "^1.11.1",
    "@keen.io/colors": "^1.1.0",
    "@keen.io/icons": "^1.5.0",
    "@keen.io/query": "^1.2.0",
    "@keen.io/typography": "^1.3.0",
    "color": "^3.1.2",
    "framer-motion": "^1.11.1",
    "fuse.js": "^6.4.1",
    "polished": "^3.4.4",
    "rc-time-picker": "^3.7.3",
    "react-color": "^2.18.1",
    "react-column-resizer": "^1.1.9",
    "react-dates": "^21.8.0",
    "react-select": "^3.1.0",
    "react-vtree": "2.0.4",
    "react-window": "^1.8.6",
    "sortablejs": "^1.13.0",
    "styled-system": "^5.1.5",
    "webfontloader": "^1.6.28"
  },
  "devDependencies": {
    "@babel/preset-env": "^7.13.8",
    "@babel/preset-typescript": "^7.13.0",
    "@quickbaseoss/babel-plugin-styled-components-css-namespace": "^1.0.1",
    "@testing-library/jest-dom": "^5.10.1",
    "@testing-library/react": "11.2.2",
    "@types/color": "^3.0.1",
    "@types/rc-time-picker": "^3.4.1",
    "@types/react-dates": "^21.8.1",
    "@types/react-select": "^3.0.13",
    "@types/react-window": "^1.8.2",
    "@types/sortablejs": "^1.10.6",
    "@types/styled-components": "^5.1.0",
    "@types/styled-system": "^5.1.9",
    "@types/webfontloader": "^1.6.29",
    "babel-plugin-styled-components": "^1.12.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
