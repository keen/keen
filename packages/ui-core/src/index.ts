import { FontLoader, Title, Text, RefText } from './typography';

import {
  Typography,
  Position,
  Alignment,
  Layout,
  ColorMode,
  RangeType,
  SortMode,
  SortByType,
  Timeframe,
  Timezones,
} from './types';

import { CDNContext } from './context';

import { screenBreakpoints } from './screen';

import {
  Anchor,
  Alert,
  Badge,
  Button,
  ActionButton,
  CircleButton,
  BulletList,
  Group,
  Card,
  CardSettings,
  Checkbox,
  ColorAdjuster,
  ContentSeparator,
  Dropdown,
  DropdownMenu,
  DropdownList,
  Error,
  FadeLoader,
  FakeInput,
  Image,
  Label,
  Loader,
  Input,
  Select,
  Modal,
  Tooltip,
  TooltipMode,
  RangeSlider,
  IntervalSlider,
  calculateIntervalValue,
  ColorScale,
  Toggle,
  Portal,
  PasswordInput,
  PasswordValidator,
  Ruler,
  RulerTick,
  Radio,
  RadioGroup,
  ARROW_SIZE,
  TagManager,
  TagManagement,
  TableCell,
  TableHeader,
  TableRow,
  CellValue,
  Brand,
  ModalContent,
  ModalHeader,
  ModalFooter,
  LoadingScreen,
  SuccessNotification,
  OAuthProviders,
  OAuthUserAction,
  OAuthConfig,
  ScrollWrapper,
  Tabs,
  AbsoluteTime,
  TIME_PICKER_CLASS,
  DatePicker,
  DropableContainer,
  DropableContainerVariant,
  RelativeTime,
  TimePeriod,
  Timezone,
  Title as TitleComponent,
} from './components';

import { UI_LAYERS, TIME_UNITS } from './constants';

export type {
  Alignment,
  CardSettings,
  ColorMode,
  TooltipMode,
  Typography,
  Position,
  Layout,
  RangeType,
  SortMode,
  SortByType,
  OAuthConfig,
  Timeframe,
  Timezones,
  DropableContainerVariant,
  CellValue,
  RulerTick,
};

export { ARROW_SIZE, UI_LAYERS, TIME_UNITS, TIME_PICKER_CLASS };

export { isElementInViewport, exportToHtml } from './utils';

export {
  Anchor,
  Alert,
  Badge,
  Button,
  ActionButton,
  CircleButton,
  BulletList,
  ColorAdjuster,
  ColorScale,
  ContentSeparator,
  Card,
  CDNContext,
  Checkbox,
  Dropdown,
  DropdownMenu,
  DropdownList,
  Error,
  FakeInput,
  FadeLoader,
  Modal,
  Title,
  Text,
  RefText,
  Group,
  Image,
  Label,
  Loader,
  FontLoader,
  Tooltip,
  Input,
  Select,
  PasswordInput,
  PasswordValidator,
  RangeSlider,
  IntervalSlider,
  calculateIntervalValue,
  Portal,
  Ruler,
  Radio,
  RadioGroup,
  TableCell,
  TableHeader,
  TableRow,
  Toggle,
  TagManager,
  TagManagement,
  Brand,
  ModalContent,
  ModalHeader,
  ModalFooter,
  LoadingScreen,
  SuccessNotification,
  screenBreakpoints,
  OAuthProviders,
  ScrollWrapper,
  Tabs,
  AbsoluteTime,
  DatePicker,
  DropableContainer,
  RelativeTime,
  TimePeriod,
  Timezone,
  TitleComponent,
  OAuthUserAction,
};
