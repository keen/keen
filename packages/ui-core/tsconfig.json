{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./lib",
    "declarationDir": "./typings",
    "target": "esnext"
  },
  "references": [
    { "path": "../colors" },
    { "path": "../icons" },
    { "path": "../query" }
  ],
  "include": ["./src/**/*"]
}
