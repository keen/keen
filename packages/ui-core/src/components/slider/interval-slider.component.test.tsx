import React from 'react';
import { mount } from 'enzyme';
import 'jest-styled-components';

import { Tooltip } from '@keen.io/ui-core';

import Rail from './rail';
import Mark from './mark';

import IntervalSlider from './interval-slider.component';

describe('@keen.io/ui-core - <RangeSlider />', () => {
  const props = {
    minimum: 0,
    maximum: 0,
    colors: ['blue', 'red'],
    intervals: [
      {
        minimum: 0,
        maximum: 100,
        step: 1,
      },
    ],
  };

  it('should show tooltip for "mouseEnter" event on drag control', () => {
    const wrapper = mount(<IntervalSlider {...props} />);
    wrapper.find(Mark).first().simulate('mouseEnter');

    expect(wrapper.find(Tooltip).length).toBeTruthy();
  });

  it('should render <Rail /> component', () => {
    const railSettings = {
      size: 10,
      borderRadius: 5,
    };

    const wrapper = mount(
      <IntervalSlider {...props} railSettings={railSettings} />
    );
    const rail = wrapper.find(Rail).first();

    expect(rail.props()).toMatchObject(railSettings);
  });

  it('should set initial value', () => {
    const wrapper = mount(<IntervalSlider {...props} initialValue={50} />);

    expect(wrapper).toMatchSnapshot();
  });
});
