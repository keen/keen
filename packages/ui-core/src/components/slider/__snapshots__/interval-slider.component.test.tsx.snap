// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@keen.io/ui-core - <RangeSlider /> should set initial value 1`] = `
.c2 {
  box-sizing: border-box;
  width: 15px;
  height: 15px;
  border-radius: 50%;
  background: #FFFFFF;
  border: solid 2px #F0AB30;
}

.c0 {
  border-radius: 3px;
  background: #E1E2E4;
}

.c1 {
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c3 {
  position: absolute;
  top: 0;
  left: 0;
  -webkit-transform: translateX(-50%) translateX(7.5px) translateY(-50%) translateY( 23px );
  -ms-transform: translateX(-50%) translateX(7.5px) translateY(-50%) translateY( 23px );
  transform: translateX(-50%) translateX(7.5px) translateY(-50%) translateY( 23px );
}

<Component
  colors={
    Array [
      "blue",
      "red",
    ]
  }
  initialValue={50}
  intervals={
    Array [
      Object {
        "maximum": 100,
        "minimum": 0,
        "step": 1,
      },
    ]
  }
  maximum={0}
  minimum={0}
>
  <div
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "height": "15px",
        "position": "relative",
      }
    }
  >
    <Rail
      borderRadius={3}
      colors={
        Array [
          "blue",
          "red",
        ]
      }
      size={4}
      type="horizontal"
    >
      <div
        style={
          Object {
            "background": "linear-gradient(90deg, blue, red",
            "borderRadius": 3,
            "height": "4px",
            "position": "absolute",
            "top": "50%",
            "transform": "translateY(-50%)",
            "width": "100%",
          }
        }
      />
    </Rail>
    <OffRange
      borderRadius={3}
      styles={
        Object {
          "height": "4px",
          "left": "0px",
          "position": "absolute",
          "top": "50%",
          "transform": "translateY(-50%)",
          "width": "calc(100% - 0px)",
        }
      }
    >
      <styled.div
        borderRadius={3}
        style={
          Object {
            "height": "4px",
            "left": "0px",
            "position": "absolute",
            "top": "50%",
            "transform": "translateY(-50%)",
            "width": "calc(100% - 0px)",
          }
        }
      >
        <div
          className="c0"
          style={
            Object {
              "height": "4px",
              "left": "0px",
              "position": "absolute",
              "top": "50%",
              "transform": "translateY(-50%)",
              "width": "calc(100% - 0px)",
            }
          }
        />
      </styled.div>
    </OffRange>
    <Component
      controlStyles={
        Object {
          "x": 0,
        }
      }
      dragConstraints={
        Object {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        }
      }
      dragDirection="x"
      onDrag={[Function]}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <ForwardRef(MotionComponent)
        drag="x"
        dragConstraints={
          Object {
            "bottom": 0,
            "left": 0,
            "right": 0,
            "top": 0,
          }
        }
        dragElastic={0}
        dragMomentum={false}
        onDrag={[Function]}
        onDragEnd={[Function]}
        onDragStart={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "display": "block",
            "position": "absolute",
            "x": 0,
            "zIndex": 1,
          }
        }
      >
        <div
          draggable={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "display": "block",
              "position": "absolute",
              "transform": "none",
              "userSelect": "none",
              "zIndex": 1,
            }
          }
        >
          <styled.div
            dragDirection="x"
          >
            <div
              className="c1"
            >
              <Mark
                backgroundColor="#FFFFFF"
                borderColor="#F0AB30"
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                size={15}
              >
                <styled.div
                  backgroundColor="#FFFFFF"
                  borderColor="#F0AB30"
                  onMouseEnter={[Function]}
                  onMouseLeave={[Function]}
                  size={15}
                >
                  <div
                    className="c2"
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    size={15}
                  />
                </styled.div>
              </Mark>
              <TooltipPosition
                containerSize={15}
                position="bottom"
                visible={false}
                x={0}
                y={0}
              >
                <styled.div
                  containerSize={15}
                  height={0}
                  position="bottom"
                  width={0}
                >
                  <div
                    className="c3"
                    height={0}
                    width={0}
                  >
                    <AnimatePresence />
                  </div>
                </styled.div>
              </TooltipPosition>
            </div>
          </styled.div>
        </div>
        <Memo(MountComponent)
          innerRef={
            Object {
              "current": .c1 {
  box-sizing: border-box;
  width: 15px;
  height: 15px;
  border-radius: 50%;
  background: #FFFFFF;
  border: solid 2px #F0AB30;
}

.c0 {
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c2 {
  position: absolute;
  top: 0;
  left: 0;
  -webkit-transform: translateX(-50%) translateX(7.5px) translateY(-50%) translateY( 23px );
  -ms-transform: translateX(-50%) translateX(7.5px) translateY(-50%) translateY( 23px );
  transform: translateX(-50%) translateX(7.5px) translateY(-50%) translateY( 23px );
}

<div
                draggable="false"
                style="display: block; position: absolute; z-index: 1; transform: none; user-select: none;"
              >
                <div
                  class="c0"
                >
                  <div
                    class="c1"
                    size="15"
                  />
                  <div
                    class="c2"
                    height="0"
                    width="0"
                  />
                </div>
              </div>,
            }
          }
          isStatic={false}
          values={
            MotionValuesMap {
              "hasMounted": true,
              "onUpdate": undefined,
              "output": [Function],
              "unsubscribers": Map {
                "x" => [Function],
              },
              "values": Map {
                "x" => MotionValue {
                  "canTrackVelocity": true,
                  "current": 0,
                  "lastUpdated": 0,
                  "notifySubscriber": [Function],
                  "parent": undefined,
                  "prev": 0,
                  "renderSubscribers": Set {
                    [Function],
                  },
                  "scheduleVelocityCheck": [Function],
                  "setChild": [Function],
                  "timeDelta": 0,
                  "transformer": undefined,
                  "updateAndNotify": [Function],
                  "updateSubscribers": Set {
                    [Function],
                  },
                  "velocityCheck": [Function],
                },
              },
            }
          }
        />
        <Component
          controls={
            ValueAnimationControls {
              "activeOverrides": Set {},
              "baseTarget": Object {
                "x": 0,
              },
              "hasValue": [Function],
              "isAnimating": Set {},
              "makeTargetAnimatable": [Function],
              "overrides": Array [],
              "props": Object {
                "children": <styled.div
                  dragDirection="x"
                >
                  <React.Fragment>
                    <Mark
                      backgroundColor="#FFFFFF"
                      borderColor="#F0AB30"
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      size={15}
                    />
                    <TooltipPosition
                      containerSize={15}
                      position="bottom"
                      visible={false}
                      x={0}
                      y={0}
                    >
                      <AnimatePresence />
                    </TooltipPosition>
                  </React.Fragment>
                </styled.div>,
                "drag": "x",
                "dragConstraints": Object {
                  "bottom": 0,
                  "left": 0,
                  "right": 0,
                  "top": 0,
                },
                "dragElastic": 0,
                "dragMomentum": false,
                "onDrag": [Function],
                "onDragEnd": [Function],
                "onDragStart": [Function],
                "onMouseDown": [Function],
                "onTouchStart": [Function],
                "style": Object {
                  "display": "block",
                  "position": "absolute",
                  "x": 0,
                  "zIndex": 1,
                },
              },
              "readValueFromSource": [Function],
              "resolvedOverrides": Array [],
              "values": MotionValuesMap {
                "hasMounted": true,
                "onUpdate": undefined,
                "output": [Function],
                "unsubscribers": Map {
                  "x" => [Function],
                },
                "values": Map {
                  "x" => MotionValue {
                    "canTrackVelocity": true,
                    "current": 0,
                    "lastUpdated": 0,
                    "notifySubscriber": [Function],
                    "parent": undefined,
                    "prev": 0,
                    "renderSubscribers": Set {
                      [Function],
                    },
                    "scheduleVelocityCheck": [Function],
                    "setChild": [Function],
                    "timeDelta": 0,
                    "transformer": undefined,
                    "updateAndNotify": [Function],
                    "updateSubscribers": Set {
                      [Function],
                    },
                    "velocityCheck": [Function],
                  },
                },
              },
              "variants": Object {},
            }
          }
          drag="x"
          dragConstraints={
            Object {
              "bottom": 0,
              "left": 0,
              "right": 0,
              "top": 0,
            }
          }
          dragElastic={0}
          dragMomentum={false}
          innerRef={
            Object {
              "current": <div
                draggable="false"
                style="display: block; position: absolute; z-index: 1; transform: none; user-select: none;"
              >
                <div
                  class="c1"
                >
                  <div
                    class="c2"
                    size="15"
                  />
                  <div
                    class="c3"
                    height="0"
                    width="0"
                  />
                </div>
              </div>,
            }
          }
          key="drag"
          onDrag={[Function]}
          onDragEnd={[Function]}
          onDragStart={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          parentContext={
            Object {
              "static": false,
            }
          }
          style={
            Object {
              "display": "block",
              "position": "absolute",
              "x": 0,
              "zIndex": 1,
            }
          }
          values={
            MotionValuesMap {
              "hasMounted": true,
              "onUpdate": undefined,
              "output": [Function],
              "unsubscribers": Map {
                "x" => [Function],
              },
              "values": Map {
                "x" => MotionValue {
                  "canTrackVelocity": true,
                  "current": 0,
                  "lastUpdated": 0,
                  "notifySubscriber": [Function],
                  "parent": undefined,
                  "prev": 0,
                  "renderSubscribers": Set {
                    [Function],
                  },
                  "scheduleVelocityCheck": [Function],
                  "setChild": [Function],
                  "timeDelta": 0,
                  "transformer": undefined,
                  "updateAndNotify": [Function],
                  "updateSubscribers": Set {
                    [Function],
                  },
                  "velocityCheck": [Function],
                },
              },
            }
          }
        />
      </ForwardRef(MotionComponent)>
    </Component>
  </div>
</Component>
`;
